src/service/viewport-impl.js

```diff
@@ -395,15 +395,11 @@ export class Viewport {
    * @return {!Promise<!../layout-rect.LayoutRectDef>}
    */
   getLayoutRectAsync(el) {
-    const scrollLeft = this.getScrollLeft();
-    const scrollTop = this.getScrollTop();
-
     // Go up the window hierarchy through friendly iframes.
     const frameElement = getParentWindowFrameElement(el, this.ampdoc.win);
     if (frameElement) {
       const bPromise = this.binding_.getLayoutRectAsync(el, 0, 0);
-      const cPromise = this.binding_.getLayoutRectAsync(
-          frameElement, scrollLeft, scrollTop);
+      const cPromise = this.binding_.getLayoutRectAsync(frameElement);
```

Comment:
/cc @dvoytenko for the not passing scroll values anymore as I am not sure what the effects of it would be.

Reply:
I found if scrollTop/scrollLeft are not passed in, binding will calculate latest scrollTop/scrollLeft. That's what I want.

Reply:
ok, please ping @dvoytenko for the implications of this before merging

Reply:
Hmm. This sounds odd. A big change from the original change. Please clarify.

Reply:
I added two new methods to `Viewport` class:
`#getLayoutRectAsync()` and `#getBoundingRectAsync()`. The idea is that in most cases where an element wants to know its layout rect, `#getLayoutRectAsync()` should be used instead of `#getLayoutRect()`. And `#getLayoutRectAsync()` always return the latest rect in certain animation frame. 

To implement this, I added a `#getLayoutRectAsync()` function to `ViewportBindingDef` class. Here I am trying to implement `Viewport` `#getLayoutRectAsync()` with `ViewportBindingDef` `#getLayoutRectAsync()` method. I am not passing scroll values here. Because I want to get the latest layout rect, and don't want to use any cached value here. (I just realized this may not work in `ViewportBindingNaturalIosEmbed_ ` case, I'll fix that) 

WDYT?

Reply:
This should definitely be using the cached values, they should _always_ be correct.

Reply:
From what I found they are not always correct when we try to calculate rect at a higher rate than browser 'scroll' event. See https://github.com/ampproject/amphtml/blob/master/src/service/viewport-impl.js#L258 that `Viewport` class use the cached `scrollTop_` value when calling `#getScrollTop()`. But `viewport.scrollTop_` is only updated on browser `scroll` event (throttled already).  https://github.com/ampproject/amphtml/blob/master/src/service/viewport-impl.js#L889.

To get the latest layout rect in each animationFrame, I can't use the cached values.

Reply:
@jridgewell @dvoytenko Maybe we can talk about this offline? My plan to to have `Viewport` provide functions to get latest rect info at every animationFrame.